Add NavigationServer random point queries

smix8 opened the pull request:
Adds query functions to get random points on navigation mesh to the NavigationServer.

Implements https://github.com/godotengine/godot-proposals/issues/6181.

Random points can be queried per navigation map or per navigation region.
Random points can be queried uniformly (costs more performance) or not.

![uniform_random_point_01](https://user-images.githubusercontent.com/52464204/226175363-ad55cef5-c071-40a8-99e9-c0dfe1be2c23.gif)


<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

AThousandShips commented:
It looks like there's some setup each time queried, would it be worth it to allow batching this, requesting several points at once?

smix8 commented:
@AThousandShips 
I am not sure. It would complicate the API to return multiple points for something that is not really performance critical and since the query allows for a navigation_layers filter caching parts of it is properly not very efficient with all the possible variants. I would guess people would also expect more or less equally spaced random positions when requesting multiple points, something that is not intended to be covered here.

smix8 committed:
Add NavigationServer random point queries

Adds query functions to get random points on navigation mesh to the NavigationServer.

smix8 commented on modules/navigation/godot_navigation_server.h:
removed

AThousandShips commented:
Will test the code when I have the time but the code looks good to me

AThousandShips reviewed (approved):
The code looks good, unable to do any real runtime testing myself

akien-mga commented:
Thanks!

tudor07 commented:
I needed this, thank you so much!

esklarski commented:
Wow wow wow ❤️

Thanks to everyone involved! This was just the thing I was looking for.

Dorifor commented:
region_get_random_point() does return me a Vector3 but it's always of value (0, `cell_height`, 0), has this happened to anyone ?

map_get_random_point() does work but only like half of the time, otherwise I also get the same value as region.

jadenrose commented:
> region_get_random_point() does return me a Vector3 but it's always of value (0, 0, 0), has this happened to anyone ?
> 
> map_get_random_point() does work but only like half of the time, otherwise I also get the same value as region.

I just ran into this issue as well! According to the [docs](https://docs.godotengine.org/en/stable/tutorials/navigation/navigation_using_navigationservers.html#waiting-for-synchronization), any procedural operations won't be ready to use until 1 physics frame after the NavigationServer is mounted, at least that's my understanding. To solve this, I deferred my call to the method that was querying the server, then at the top of that method, I waited for a `get_tree().physics_frame` signal. After that, the queries started to work consistently.

Example:
```
  func on_enter():
-   goto_random_destination()
+   goto_random_destination.call_deferred()

...

  func goto_random_destination():
+   await get_tree().physics_frame

    # ... get your map
    var random_point = NavigationServer3D.map_get_random_point(map, 1, false)
```
